Sprint 4

vladkramskoy opened the pull request:
Получилось сделать:
- Экран, верстка
Были трудности:
- Экземпляры структур моделей
- Использование self
- Правильное расположение блоков кода, отступы от блоков и строк
- Объявление переменных и функции: в классе MovieQuizViewController: UIViewController { ... } (что располагать в теле класса, а что после него)

vladkramskoy committed:
Добавил ассеты, сверстал первую кнопку

vladkramskoy committed:
Настроен launch screen

vladkramskoy committed:
Реализован главный экран приложения

vladkramskoy committed:
Созданы mock-данные, созданы необходимые аутлеты и экшены

vladkramskoy committed:
Реализована первичная логика приложения:
- Стартовый вопрос при запуске
- Переключение вопросов после ответа пользователя
- Результат ответа в виде зеленой или красной рамки
- Окно с результатом викторины, кол-вом верных ответов и возможностью начать сначала

Nikitaaeee commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,161 @@
-    // MARK: - Lifecycle
+    
+    // Аутлеты и экшены
+    @IBOutlet private weak var imageView: UIImageView!
+    @IBOutlet private weak var textLabel: UILabel!
+    @IBOutlet private weak var counterLabel: UILabel!
```
Круто, что не забываешь добавлять `weak` для `@IBOutlet`! 👍🏻

vladkramskoy replied:
👍🏻

Nikitaaeee commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,161 @@
+    @IBOutlet private weak var textLabel: UILabel!
+    @IBOutlet private weak var counterLabel: UILabel!
+    
+    @IBAction private func yesButtonClicked(_ sender: UIButton) {
+        let currentQuestion = questions[currentQuestionIndex]
+        let givenAnswer = true
+        showAnswerResult(isCorrect: givenAnswer == currentQuestion.correctAnswer)
+    }
+    @IBAction private func noButtonClicked(_ sender: UIButton) {
+        let currentQuestion = questions[currentQuestionIndex]
+        let givenAnswer = false
+        showAnswerResult(isCorrect: givenAnswer == currentQuestion.correctAnswer)
+    }
+    
+    // Показываем первый экран
+    private func show(quiz step: QuizStepViewModel) {
+        imageView.image = step.image
+        textLabel.text = step.question
+        counterLabel.text = step.questionNumber
+    }
+    
```
Обычно используем такой порядок: 
- сначала перечисляем свойства класса, 
- затем функции. 
 
При этом также делим на приватные и публичные блоки. 
Функции перечисляют в таком порядке: 
- `init`ы, 
- `override`ы, 
- блоки публичных и приватных функций,
- блок c аннотацией `@IBAction`

vladkramskoy replied:
👍🏻

Nikitaaeee commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,161 @@
+    
+    // Если вопрос не последний, то показываем следующий
+    private func showQuestionOrResults() {
+        // Блок показывает результат
+        if currentQuestionIndex == questions.count - 1 {
+            let text = "Ваш результат: \(correctAnswers)/\(questions.count)"
+            let resultAlert = QuizResultsViewModel(title: "Этот раунд окончен!", text: text, buttonText: "Сыграть ещё раз")
+            showResult(quiz: resultAlert)
+        } else {
+            // Блок показывает след. вопрос
+            currentQuestionIndex += 1
+            let nextQuestion = questions[currentQuestionIndex]
+            let viewModel = convert(model: nextQuestion)
+            show(quiz: viewModel)
+        }
+    }
+    
+    // Блок с алертом
+    private func showResult(quiz result: QuizResultsViewModel) {
+        let alert = UIAlertController(
+            title: "Этот раунд окончен!",
+            message: "Ваш результат: \(correctAnswers)/\(questions.count)",
+            preferredStyle: .alert)
+        
+        let action = UIAlertAction(title: "Сыграть ещё раз", style: .default) { _ in
+            currentQuestionIndex = 0
+            correctAnswers = 0
+            
+            let firstQuestion = questions[currentQuestionIndex]
+            let viewModel = convert(model: firstQuestion)
+            self.show(quiz: viewModel)
+        }
+        
+        alert.addAction(action)
+        
+        self.present(alert, animated: true, completion: nil)
+    }
+    
+    // Результат ответа, зеленая/красная рамка
+    private func showAnswerResult(isCorrect: Bool) {
```
Комментарии могут быть очень полезными, и я заметил, что ты прилагаешь усилия к тому, чтобы добавлять пояснения к некоторым методам/свойствам.

Однако, я хотел бы предложить более умеренный подход к комментированию. 
Хорошо структурированный и читаемый код часто говорит сам за себя. Если название метода или свойства ясно отражает его назначение и его поведение легко понять из контекста, то комментарий может быть не нужен.

Эта тема хорошо разобрана в книге "Чистый код" Роберта Мартина.
Также про комментарии в коде можно посмотреть тут:
https://habr.com/ru/companies/enterra/articles/182574/

На этапе обучения оставлять комментарии - нормально, так как они могут помогать ориентироваться в коде. Но по мере обучения лучше стараться убирать комментарии такого вида:)

При этом не все комментарии принято удалять
Для организации кода используют комментарии такого типа:
`// MARK: - Private Methods`
Про организацию кода можно посмотреть тут:
https://github.com/RedMadRobot/RMR-swift-style-guide#code-organization

vladkramskoy replied:
Спасибо, буду работать над этим, буду постепенно сводить их к минимуму

Nikitaaeee commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,161 @@
+private var currentQuestionIndex = 0
+private var correctAnswers = 0
+
+// Модели
+struct QuizQuestion {
+    let image: String
+    let text: String
+    let correctAnswer: Bool
+}
+struct QuizStepViewModel {
+    let image: UIImage
+    let question: String
+    let questionNumber: String
+}
+struct QuizResultsViewModel {
+    let title: String
+    let text: String
+    let buttonText: String
+}
```
Модели можно вынести в отдельный файл.
Для разработчика может быть не очень удобно и неочевидно листать файл с контроллером вниз, чтобы найти там структуры моделей 🙂

vladkramskoy replied:
👍🏻

Nikitaaeee commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,161 @@
+        correctAnswer: true),
+    QuizQuestion(
+        image: "The Dark Knight",
+        text: "Рейтинг этого фильма больше чем 6?",
+        correctAnswer: true),
+    QuizQuestion(
+        image: "Kill Bill",
+        text: "Рейтинг этого фильма больше чем 6?",
+        correctAnswer: true),
+    QuizQuestion(
+        image: "The Avengers",
+        text: "Рейтинг этого фильма больше чем 6?",
+        correctAnswer: true),
+    QuizQuestion(
+        image: "Deadpool",
+        text: "Рейтинг этого фильма больше чем 6?",
+        correctAnswer: true),
+    QuizQuestion(
+        image: "The Green Knight",
+        text: "Рейтинг этого фильма больше чем 6?",
+        correctAnswer: true),
+    QuizQuestion(
+        image: "Old",
+        text: "Рейтинг этого фильма больше чем 6?",
+        correctAnswer: false),
+    QuizQuestion(
+        image: "The Ice Age Adventures of Buck Wild",
+        text: "Рейтинг этого фильма больше чем 6?",
+        correctAnswer: false),
+    QuizQuestion(
+        image: "Tesla",
+        text: "Рейтинг этого фильма больше чем 6?",
+        correctAnswer: false),
+    QuizQuestion(
+        image: "Vivarium",
+        text: "Рейтинг этого фильма больше чем 6?",
+        correctAnswer: false)
+]
+private var currentQuestionIndex = 0
+private var correctAnswers = 0
```
Свойства принято хранить внутри объектов, которые их используют.
Сейчас эти свойства лежат в файле с MovieQuizViewController, но при этом вне области самого класса.
Можно перенести их в класс.

vladkramskoy replied:
Спасибо, ваши комментарии очень помогли)

vladkramskoy committed:
Исправлены недочеты:
- Исправлена рамка индикации ответов
- Исправлено множественное нажатие кнопок Да/Нет
- Улучшена структура кода

Nikitaaeee commented:
👍🏻👍🏻👍🏻
